Vouchers: sort matching courseruns by similarity

umarmughal824 opened the pull request:
#### Pre-Flight checklist

- [x] Screenshots and design review for any changes that affect layout or styling
  - [x] Desktop screenshots
  - [x] Mobile width screenshots
  - [x] Tag @ferdi or @pdpinch for review
- [x] Testing
  - [x] Code is tested
  - [x] Changes have been manually tested


#### What's this PR do?
fixes #1227 

#### How should this be manually tested?
You need to perform the following steps for generating seed data of vouchers.
**Add Environment Variables**

> VOUCHER_DOMESTIC_DATE_KEY=UNIQUE01
VOUCHER_DOMESTIC_EMPLOYEE_KEY=UNIQUE02
VOUCHER_DOMESTIC_EMPLOYEE_ID_KEY=UNIQUE03
VOUCHER_DOMESTIC_KEY=UNIQUE04
VOUCHER_DOMESTIC_COURSE_KEY=UNIQUE05
VOUCHER_DOMESTIC_CREDITS_KEY=UNIQUE06
VOUCHER_DOMESTIC_DATES_KEY=UNIQUE07
VOUCHER_DOMESTIC_AMOUNT_KEY=UNIQUE08
VOUCHER_INTERNATIONAL_EMPLOYEE_KEY=UNIQUE09
VOUCHER_INTERNATIONAL_PROGRAM_KEY=UNIQUE10
VOUCHER_INTERNATIONAL_COURSE_KEY=UNIQUE11
VOUCHER_INTERNATIONAL_SCHOOL_KEY=UNIQUE12
VOUCHER_INTERNATIONAL_EMPLOYEE_ID_KEY=UNIQUE13
VOUCHER_INTERNATIONAL_AMOUNT_KEY=UNIQUE14
VOUCHER_INTERNATIONAL_DATES_KEY=UNIQUE15
VOUCHER_INTERNATIONAL_COURSE_NAME_KEY=UNIQUE16
VOUCHER_INTERNATIONAL_COURSE_NUMBER_KEY=UNIQUE17

**1. Create Course with these details**

course readable_id: AMxB
course name: Additive Manufacturing for Innovative Design and Production
dates: 04/30/2018 - 07/02/2018

![test_voucher](https://user-images.githubusercontent.com/4043989/67491963-28656280-f68f-11e9-841c-5516b5031ed0.png)


**2. Create Course Run**

![test_voucher1](https://user-images.githubusercontent.com/4043989/67491970-2b605300-f68f-11e9-8698-d1fcbdd99f7a.png)


**3. Create Product**

![test_voucher2](https://user-images.githubusercontent.com/4043989/67492344-b8a3a780-f68f-11e9-99d8-52efc5321572.png)


**4. Create Product Version**

![test_voucher3](https://user-images.githubusercontent.com/4043989/67492353-bb9e9800-f68f-11e9-81b8-259dbf34643f.png)


**5. Create Coupon with 100% discount in Company id = 1**

![test_voucher4](https://user-images.githubusercontent.com/4043989/67492492-ee489080-f68f-11e9-884d-f7eccd12114e.png)

**6. Try to upload ```mitxpro/voucher/.test/domestic_voucher.pdf``` on ```xpro.odl.local:port/boeing/upload/``` you will get one course on ```xpro.odl.local:port/boeing/enroll/```.**


**7. You need to create multiple Course Runs for this Course for multiple values in the dropdown but with a different name and readable_id but keep the course run date same ```(2018-04-30)``` .**

**Flow**
Course Run -> Product -> Product Version -> Create Coupon
 


#### Screenshots (if appropriate)
**Before sorting**
![before_sorting](https://user-images.githubusercontent.com/4043989/67493476-5d72b480-f691-11e9-8465-bcf33eed927d.png)

**After sorting**
![after_sorting](https://user-images.githubusercontent.com/4043989/67493489-606da500-f691-11e9-8c28-cfe47161d373.png)

pdpinch commented on voucher/utils.py:
```diff
@@ -113,6 +114,22 @@ def get_eligible_coupon_choices(voucher):
+            eligible_choices_titles,
+            len(eligible_choices_titles),
+            0,
+        )
```
will get_close_matches always return one item?

umarmughal824 replied:
It will return all of the items in sorted orders with respect to the close match. The best-matched item will come first and no matched item in the end.

pdpinch commented on voucher/templates/enroll.html:
```diff
@@ -13,6 +13,7 @@ <h2>Confirm Course</h2>
             <form method="post" enctype="multipart/form-data" class="voucher-form ">
                 {% csrf_token %}
                 <select name="coupon_version" id="coupon_version" classname="form-control">
+                    <option value="">Choose Course</option>
```
What happens if the user tries to submit the form when this option is still chosen?

umarmughal824 replied:
Currently its giving error. I will address it too in my PR.

pdpinch commented:
@umarmughal824 if this is ready for review, please add the flag. I think someone from @ahmed-arbisoft should be able to do the full review.

umarmughal824 commented:
@pdpinch I am working on its test cases. Once I will be done, I will mark it for review.

umarmughal824 commented:
@pdpinch the user can't enroll with default option **Choose Course**. This is how it looks like If I try to submit it with the default option.

![compulsory](https://user-images.githubusercontent.com/4043989/67679708-83f36100-f9ab-11e9-82b6-8a3b91030d55.png)

pdpinch commented:
Good. It is correct that the user can’t enroll. 

However, this is not our standard design for an error message. Where is that styling coming from?

umarmughal824 commented:
@pdpinch Its HTML5 builtin form validation error message, I have marked that field required. Please let me know if we need to change it.

pdpinch commented:
Do we do that any where else on the site?

Is the form being validated on the back end?

ahmed-arbisoft commented:
I guess we need to update the form to use our own validation and error messages. Custom styling is being used in both Django- and React-rendered pages.

Also, I guess take another look at the server-side validation.
On Oct 28 2019, at 6:22 pm, Peter Pinch <notifications@github.com> wrote:
> Do we do that any where else on the site?
>
> Is the form being validated on the back end?
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub (https://link.getmailspring.com/link/F8088498-A62C-4B37-8809-1F5DB092F9F1@getmailspring.com/0?redirect=https%3A%2F%2Fgithub.com%2Fmitodl%2Fmitxpro%2Fpull%2F1257%3Femail_source%3Dnotifications%26email_token%3DAKZ74EQNXDYZK3HCIMYEQM3QQ3RRFA5CNFSM4JEUFC22YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECM3A6Q%23issuecomment-546943098&recipient=cmVwbHkrQUtaNzRFWFYyVkhDWlpJT0dMM09NSVYzWVFQQkZFVkJOSEhCNUE1N0NZQHJlcGx5LmdpdGh1Yi5jb20%3D), or unsubscribe (https://link.getmailspring.com/link/F8088498-A62C-4B37-8809-1F5DB092F9F1@getmailspring.com/1?redirect=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAKZ74ERHQT3CND7WCL3552LQQ3RRFANCNFSM4JEUFC2Q&recipient=cmVwbHkrQUtaNzRFWFYyVkhDWlpJT0dMM09NSVYzWVFQQkZFVkJOSEhCNUE1N0NZQHJlcGx5LmdpdGh1Yi5jb20%3D).
>
>


-- 


The content of this email and attachment(s) are confidential and intended 
solely for the recipient(s) specified in the message. If you received this 
email by error, please alert the sender by replying to this email and 
follow with its deletion, so that we can ensure such an error does not 
occur in future. If you are not the intended recipient, you are strictly 
prohibited to use, copy, disseminate or store this email or its 
attachment(s).

umarmughal824 commented:
@pdpinch I have updated the error message identical to the rest of the website and added backend validation too. This is how it looks like now.
![look](https://user-images.githubusercontent.com/4043989/67773717-f92f6680-fa7d-11e9-8a66-a4ddf28eb5eb.png)

pdpinch commented on voucher/templates/enroll.html:
```diff
@@ -10,18 +10,46 @@ <h1>Boeing employee enrollment page</h1>
+                            <option value="{{ choice.0 }}">{{ choice.1 }}</option>
+                        {% endfor %}
+                    </select>
+                    <div id="voucher-error" class="form-error d-none">Please select an item in the list.</div>
```
```suggestion
                    <div id="voucher-error" class="form-error d-none">Please select a course from the list. If your expected course isn't in the list, please contact <a href="https://xpro.zendesk.com/hc/en-us/requests/new" target="_blank" rel="noopener noreferrer">Customer Support</a></div>
```

Hopefully my HTML is correct...

umarmughal824 commented:
@pdpinch I have updated the error message and this is how looks now. 
![look](https://user-images.githubusercontent.com/4043989/67775192-6512ce80-fa80-11e9-8603-e8624705113e.png)

pdpinch commented on voucher/templates/enroll.html:
```diff
@@ -10,18 +10,46 @@ <h1>Boeing employee enrollment page</h1>
+                            <option value="{{ choice.0 }}">{{ choice.1 }}</option>
+                        {% endfor %}
+                    </select>
+                    <div id="voucher-error" class="form-error d-none">Please select a course from the list. If your expected course isn't in the list, please contact <a href="https://xpro.zendesk.com/hc/en-us/requests/new" target="_blank" rel="noopener noreferrer">Customer Support</a></div>
```
a couple of nitpicks:

1. I should have put a period at the end of this sentence:

2. My length of this line I suggested is way too long. Does the code linter complain about this? Please add at least one line break. 

```suggestion
                    <div id="voucher-error" class="form-error d-none">Please select a course from the list. 
                        If your expected course isn't in the list, please contact 
                        <a href="https://xpro.zendesk.com/hc/en-us/requests/new" target="_blank" rel="noopener noreferrer">Customer Support</a>.
                    </div>
```

umarmughal824 replied:
Our code linter doesn't work for HTML.

umarmughal824 commented:
@pdpinch this is how it looks like now. 
![updated](https://user-images.githubusercontent.com/4043989/68108196-f9f64b80-ff08-11e9-9e64-116114e00ec9.png)
